src/utils/RequestQueue.ts

```diff
@@ -0,0 +1,256 @@
+      // If a delay is set, wait to add this to the queue.
+      if (delayMs > 0) {
+        const timeoutId = setTimeout(() => this.addRequestToQueue(key), delayMs);
+        // Save timeout information to request metadata
+        requestItem.timeoutId = timeoutId;      
+      } else {
+        // No delay, add immediately
+        this.addRequestToQueue(key);
+      }
+    } else if (delayMs <= 0) {
+      // This request is registered, but is now being requested without a delay.
+      // Move into queue immediately if it's not already added, and clear any timeouts it may have.
+      this.addRequestToQueue(key);
+    }
+
+    const promise = this.allRequests.get(key)?.promise;
+    if (!promise) {
+      throw new Error("Found no promise to return when getting stored request data.");
+    }
+    return promise;
+  }
+
+  /**
+   * Adds multiple requests to the queue, with an optional delay between each.
+   * @param requests An array of RequestItems, which include a key and a request action.
+   * @param delayMs An optional minimum delay in milliseconds to be added between each request.
+   *  For example, a delay of 10 ms will cause the second request to be added to the processing queue
+   *  after 10 ms, the third to added after 20 ms, and so on. Set to 10 ms by default.
+   * @returns An array of promises corresponding to the provided requests. (i.e., the `i`th value
+   * of the returned array will be a Promise for the resolution of `requests[i]`). If a request
+   *  with a matching key is already pending, returns the promise for the initial request.
+   */
+  public addRequests<T>(requests: Request<T>[], delayMs = 10): Promise<unknown>[] {
+    const promises: Promise<unknown>[] = [];
+    for (let i = 0; i < requests.length; i++) {
+      const item = requests[i];
+      const promise = this.addRequest(item.key, item.requestAction, delayMs * i);
+      promises.push(promise);
+    }
+    return promises;
+  }
+
+  /**
+   * Attempts to remove and run the next queued request item, if resources are available.
+   * @returns true if a request was started, or false if there are too many
+   * requests already active.
+   */
+  private async dequeue(): Promise<void> {
+    if (this.activeRequests.size >= this.maxActiveRequests || this.queue.length === 0) {
+      return;
+    }
+    const requestKey = this.queue.shift();
+    if (!requestKey) {
+      return;
+    }
+    if (this.activeRequests.has(requestKey)) {
+      // This request is already active, so skip.
+      return;
+    }
+
+    const requestItem = this.allRequests.get(requestKey);
+    if (!requestItem) {
+      return;
+    }
+
+    const key = requestItem.key;
+    // Mark that this request is active
+    this.activeRequests.add(key);
+    
+    await requestItem.action().then(
+      (fulfilledValue) => {
+        requestItem.resolve(fulfilledValue);
+      }, (rejectedValue) => {
+        requestItem.reject(rejectedValue);
+      }
+    );
```

Comment:
No need to wrap `resolve`/`reject` in arrow functions here, they can be passed directly: `.then(requestItem.resolve, requestItem.reject)`

Reply:
Thanks for the tip!